src/pybind/mgr/dashboard/frontend/src/app/shared/directives/auth-storage.directive.ts

```diff
@@ -0,0 +1,74 @@
+import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import _ from 'lodash';
+
+import { Permissions } from '~/app/shared/models/permissions';
+import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
+
+@Directive({
+  selector: '[cdScope]'
+})
+export class AuthStorageDirective {
+  permissions: Permissions;
+  matchAll = false;
+
+  constructor(
+    private templateRef: TemplateRef<any>,
+    private viewContainer: ViewContainerRef,
+    private authStorageService: AuthStorageService
+  ) {}
+
+  @Input('cdScope') set cdScope(condition: string | string[] | object) {
+    this.permissions = this.authStorageService.getPermissions();
+    if (this.isAuthorized(condition)) {
+      this.viewContainer.createEmbeddedView(this.templateRef);
+    } else {
+      this.viewContainer.clear();
+    }
+  }
+
+  @Input('cdScopeMatchAll') set cdScopeMatchAll(condition: boolean) {
+    this.matchAll = condition;
+  }
+
+  private isAuthorized(condition: any): boolean {
+    /*
+      @param string
+      single permission equivalent to check e.g osd.
+
+      @param string[]
+      array permission equivalent to check e.g ["osd", "rgw"].
+
+      @param object
+      object permission equivalent to check e.g {osd: ["read", "update"], rgw: ["read", "create"]}.
+
+      @param matchAll
+      matchAll determines iwhether all permissions are required or not
+
+      We check if the permission key exists on the Permissions class then
+      if true We check for read permission and display the component accordingly
+    */
+
+    if (_.isString(condition)) {
+      return _.get(this.permissions, [condition, 'read'], false);
+    } else if (_.isArray(condition)) {
+      if (this.matchAll) {
+        return _.every(condition, (permission) => this.permissions[permission]['read']);
+      } else {
+        return _.some(condition, (permission) => this.permissions[permission]['read']);
+      }
+    } else if (_.isObject(condition)) {
+      if (this.matchAll) {
+        return _.every(condition, (value, key) => {
+          return _.every(value, (val) => this.permissions[key][val]);
+        });
+      } else {
+        return _.some(condition, (value, key) => {
+          return _.some(value, (val) => this.permissions[key][val]);
+        });
+      }
+    }
```

Comment:
A trick here: instead of branching for each matchAll value, why not replacing the function itself?:

```suggestion
    everyOrSome = this.matchAll ? _.every : _.some;
    ...
    } else if (_.isArray(condition)) {
      return everyOrSome(condition, (permission) => this.permissions[permission]['read']);
    } else if (_.isObject(condition)) {
    return everyOrSome(condition, (value, key) => {
      return everyOrSome(value, (val) => this.permissions[key][val]);
      });
    }
```

Reply:
Great! I didn't even come close to thinking of this. I'll refactor